core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -482,19 +492,29 @@ public void stopPartitions(Set<StopPartition> stopPartitions,
                         task.cancel();
                         return null;
                     });
-                    leaderExpirationRLMTasks.computeIfPresent(tpId, (topicIdPartition, task) -> {
-                        LOGGER.info("Cancelling the expiration RLM task for tpId: {}", tpId);
-                        task.cancel();
-                        return null;
-                    });
                     followerRLMTasks.computeIfPresent(tpId, (topicIdPartition, task) -> {
                         LOGGER.info("Cancelling the follower RLM task for tpId: {}", tpId);
                         task.cancel();
                         return null;
                     });
 
+                    // here, we have to handle retain and delete separately.
+                    // If "retain" is set, we should not cancel expiration task
+                    if (!REMOTE_LOG_DISABLE_POLICY_RETAIN.equals(remoteLogDisablePolicy)) {
+                        leaderExpirationRLMTasks.computeIfPresent(tpId, (topicIdPartition, task) -> {
+                            LOGGER.info("Cancelling the expiration RLM task for tpId: {}", tpId);
+                            task.cancel();
+                            return null;
+                        });
+                    } else if (REMOTE_LOG_DISABLE_POLICY_DELETE.equals(remoteLogDisablePolicy)) {
+                        // update start offset to local log start offset
+                        // we use unused "-2" as the flag to set local log start offset
+                        updateRemoteLogStartOffset.accept(tp, FLAG_TO_SET_LOCAL_LOG_START_OFFSET);
+                    }
+
                     removeRemoteTopicPartitionMetrics(tpId);
 
+                    // we already set it correctly, just apply it.
```

Comment:
We also have to update the L531-L539 to stop the RLMM task when the remote storage is disabled with "delete" policy but the topic is not deleted.

We can introduce a new field in StopPartition (stopRLMM) to decide when to stop the RLMM

Reply:
Since we'll stop RLMM when "delete" is set, I just use "delete" as an indication to stop RLMM, without introducing a new field.